arangod/RocksDBEngine/RocksDBIndex.cpp

```diff
@@ -179,22 +181,26 @@ void RocksDBIndex::setupCache() {
   // by _cacheEnabled already.
   TRI_ASSERT(!ServerState::instance()->isCoordinator());
 
-  if (_cache == nullptr) {
+  auto cache = _cache;
+  if (cache == nullptr) {
     TRI_ASSERT(_cacheManager != nullptr);
     LOG_TOPIC("49e6c", DEBUG, Logger::CACHE) << "Creating index cache";
     // virtual call!
-    _cache = makeCache();
+    cache = makeCache();
+    std::atomic_store_explicit(&_cache, std::move(cache),
+                               std::memory_order_relaxed);
   }
 }
 
 void RocksDBIndex::destroyCache() noexcept {
-  if (_cache != nullptr) {
+  auto cache = _cache;
+  if (cache != nullptr) {
     try {
       TRI_ASSERT(_cacheManager != nullptr);
       // must have a cache...
       LOG_TOPIC("b5d85", DEBUG, Logger::CACHE) << "Destroying index cache";
-      _cacheManager->destroyCache(std::move(_cache));
-      _cache.reset();
+      _cacheManager->destroyCache(std::move(cache));
+      std::atomic_store_explicit(&_cache, {}, std::memory_order_relaxed);
```

Comment:
Is it safe order, sorry thinks about it only now.
Is cache (shared_ptr with it) valid after/in process destroyCache?
Maybe should be opposite order?
Cache became invisible (for useCache) and then destroyed?

Reply:
Changed